broker/src/main/java/io/camunda/zeebe/broker/system/partitions/ZeebePartitionAdminAccess.java

```diff
@@ -140,4 +149,74 @@ public ActorFuture<Void> resumeProcessing() {
         });
     return completed;
   }
+
+  @Override
+  public ActorFuture<Void> banInstance(final long processInstanceKey) {
+    final ActorFuture<Void> future = concurrencyControl.createFuture();
+    concurrencyControl.run(
+        () -> {
+          try {
+
+            final LogStream logStream = adminControl.getLogStream();
+            logStream
+                .newLogStreamWriter()
+                .onComplete(
+                    (writer, error) -> {
+                      if (error != null) {
+                        LOG.error(
+                            "Could not retrieve writer to write error record for process instance.",
+                            error);
+                        future.completeExceptionally(error);
+                      } else {
+                        final var errorRecord = new ErrorRecord();
+                        errorRecord.initErrorRecord(
+                            new Exception("Instance was banned from outside."), -1);
+                        errorRecord.setProcessInstanceKey(processInstanceKey);
+
+                        final var recordMetadata =
+                            new RecordMetadata()
+                                .recordType(RecordType.EVENT)
+                                .valueType(ValueType.ERROR)
+                                .intent(ErrorIntent.CREATED)
+                                .recordVersion(RecordMetadata.DEFAULT_RECORD_VERSION)
+                                .rejectionType(RejectionType.NULL_VAL)
+                                .rejectionReason("");
+                        final var entry =
+                            RecordBatchEntry.createEntry(
+                                processInstanceKey, recordMetadata, -1, errorRecord);
+                        final var eitherWrittenOrFailure = writer.tryWrite(entry);
+
+                        eitherWrittenOrFailure.ifRightOrLeft(
+                            (position) -> {
+                              LOG.info("Wrote error record on position {}", position);
+
+                              // we only want to make the state change after we wrote the event
+                              final var zeebeDb = adminControl.getZeebeDb();
+                              final var context = zeebeDb.createContext();
+                              final var dbBannedInstanceState =
+                                  new DbBannedInstanceState(zeebeDb, context, partitionId);
+
+                              dbBannedInstanceState.banProcessInstance(processInstanceKey);
```

Comment:
❓ Why not use the event applying state writer instead of modifying the state directly?

I'm asking this because that is the idea behind correction records, i.e. a special type of event that describes that some corrupt data is corrected

Reply:
I tried to keep it simply and the changes were more incrementely so no particular reason. Not sure regarding the setup of the state writer since it likely needs some more stuff.